### ISSUE TITLE
Yunit in Archlinux

### ISSUE THREAD

**Issue by horsemanoffaith** [yunit-io/yunit#70]
While I am no programmer, I love Unity (I have been using Unity since its inception with Ubuntu 11.04) and would love to see Yunit in Arch Linux. I have done some work maintaining Unity-for-Arch, and would be willing to do what I can to support Yunit in it's development specific to Arch Linux. Is there any possibility that some of the packages used for the Unity-for-Arch project help in your development? Are any of those packages dependencies for Unity 8/Yunit? The Unity-for-Arch repos are located here:

https://github.com/chenxiaolong/Unity-for-Arch
https://github.com/chenxiaolong/Unity-for-Arch-Extra
https://github.com/horsemanoffaith/Unity-for-Arch

Is Yunit only specific to Unity8, or are there plans to do both Unity7 and 8?

**Comment by jsalatas** [comment]
Do you want us to invest time to review your work now or can it wait for some later time (ie after the mir to wayland transition)?

**Comment by Leppan** [comment]
Well, what similarities do you see?

**Comment by horsemanoffaith** [comment]
@jsalatas Up to you. Unity-for-Arch is built using vanilla Arch, which is still basically using X. If I have a list of the dependencies for Yunit, there is a possibility I could create the additional necessary PKGBUILDS to get Yunit working on Arch Linux. I have Unity 7 (mostly) working on both my laptop and desktop now. I know Unity 8 is a different interface with different dependencies, but if there's anything that I could do to help speed development, I'll do what I can. If you'd be willing just to look at the package list and tell me what packages are necessary for Yunit/Unity 8, that would give me at least something to start with.

@Leppan- I am not really familiar with Unity 8. I was never active in the port of Unity 8 to Arch Linux. Chenxiaolong did have a Unity 8 repo on github, but it says that all the Unity 8 files are in the AUR. AFAIK, I haven't seen any Unity 8 files on the AUR.

**Comment by horsemanoffaith** [comment]
@Leppan- I am checking on the dependencies of Unity8 on Launchpad now. I already see a couple of things that I may be able to use the existing PKGBUILDS for:

gsettings-ubuntu-schemas
libgeonames
gsettings-qt
lightdm
unity-api

This is all I see from the Unity-for-Arch package list that are direct dependencies for Unity 8 (based off the dependency list on Launchpad). I'll do more research and find out what other dependencies I'll need from there. Mir is already a package in the AUR for Arch. I'm not sure if the package actually works, but there is a REMOTE possibility I could get Unity 8 working on Arch.

**Comment by jsalatas** [comment]
@horsemanoffaith  please check the build.sh script. before trying to build things, it makes sure that all deps are met (in an ubuntu 17.04 env).

**Comment by horsemanoffaith** [comment]
I'll actually look at installing 17.04 so I can do so. What I'm doing right now is trying to build the required packages in Arch (pulling them directly from Launchpad) to see if there's any way possible I can figure out how to make unity8 run in Arch. I will probably refer to your work when needed- I'll probably need it here and there, but if what I'm doing will help you in any way, I'll be happy to contribute. I think Arch would be a great test bed for Yunit because it's systemd based and is VERY customizable.

**Comment by bregma** [comment]
Not sure if it's any help, but [here is a partial diagram](http://i.imgur.com/3e4Yq0f.jpg) of the dependencies of the Unity (8) Desktop Environment.  It was autogenerated as a guide to landing Ubuntu Touch packages in Ubuntu in preparation for having the Unity DE on the 16.10 ISO, so it doesn't show stuff that was already in Ubuntu, and after it was made some dependencies were made optional, but it gives you an idea.

If you're planning on porting Yunit to run on the GNOME DE (like Unity 7 does), a lot of this stuff will need to be replaced with GNOME equivalents, or the functionality just removed from the code.

**Comment by horsemanoffaith** [comment]
Sheeshamoly that diagram is complicated! I'm not saying all that stuff's not needed, but yikes! What I'm doing at this point is building the packages that Unity 8 requires based on the Launchpad page and kinda reverse engineering it. I think that's what Chenxiaolong did when he ported Unity 7 to Arch. I don't know that I have the know-how to do that, but so far it's been fun creating my own PKGBUILDS in Arch and seeing if they compile. So far they have, but honestly I haven't gotten that far in the process. As of this moment I'm in the process of creating an ubuntu-app-launch PKGBUILD, because the unity 8 package complains about needing it. We'll see how far I get in my endeavor- I'll probably need to get a working version of Ubuntu installed on one of my PC's and see how stuff is wired in there- that'll help me get it up and running in Arch.

**Comment by bregma** [comment]
If it's any help, the script I used to generate that diagram is [here](http://bazaar.launchpad.net/~bregma/+junk/MIR-tools/files).  It traverses the APT package dependency tree to determine all the transitive source and binary dependencies that were not in main (part of the Main Inclusion Request [MIR] process in Ubuntu).   It should be easy enough to tweak to list all packages, and not just those not in the main archives, or to do the equivalent with whatever Arch uses.

**Comment by konfou** [comment]
Resetting removal commits, @Chenxiaolong' repo had PKGBUILDs for the following packages:
```
* gflags
* liburcu
* lttng-ust
* umockdev
* mir
* mesa-mir
* xorg-server-mir
* xf86-video-intel-mir
* xf86-video-nouveau-mir
* tslib
* sdl2-mir
* unity-system-compositor
```
As said all those are on AUR though some of them lack a maintainer. Similar to Arch's Unity (v7) metapackage there is @shiznix [Gentoo's overlay](/shiznix/unity-gentoo) which includes ebuilds for Unity 8.

**Comment by horsemanoffaith** [comment]
Is there anything I can do to help you guys with this? I've already made a little progress on creating PKGBUILDS necessary for the unity8 package off Launchpad. I have seen some of the above listed packages in the AUR, and I've installed a few while I've been working on ubuntu-app-launch. My guess is getting Unity 8 working on Arch will require more than those few packages.

**Comment by shiznix** [comment]
Good luck @horsemanoffaith 
Was drawn here by way of notification on @konfou 's comment.

I pretty much got all of Unity8 desktop on Gentoo working on the same day as Canonical's announcement came through that it was dead.

Gentoo's ebuilds are extremely similar to Arch's pkgbuilds (Gentoo being an ancestor of Arch) so you might have some help gaining package dependency info from the Gentoo overlay ebuilds.
That same similarity has come in handy for both @Chenxiaolong 's and the unity-gentoo overlay in the past when we were both attempting in parallel to get Unity onto a non-Ubuntu system.

Some other things to think about for Unity 8's runtime:
* You will be required to use Ubuntu's QT5 qtbase patchset for all core QT packages
* Your systemd enabled system and kernel will also require a working lxc,lxcfs,cgroup,cgmanager setup
* You will have to apply and maintain some hacks around various Poettering systemd decisions wrt. stifling cgroups and cgmanager
* Upstart is required to not only start desktop session services but in a Unity8 world is also responsible for launching every application

So anyway, have at it!
I'm tending to be heading down the road of dropping Unity8 from the overlay as some of it's fragile design choices just seemed kind of poor from the beginning.

Cheers :)

**Comment by bregma** [comment]
Ubuntu forwards all of Debian's patches to the qtbase package and add a few more  for things like tweaking appmenus.  None of those patches are specific to Unity 8 or even required by Unity 8.

There is nothing in the Unity 8 stack that requires LXC or LXCFS.  You are definitely going to need a modern kernel that supports cgroups.  They're been around for years, though, so that shouldn't be a problem.

Upstart is *only* required for launching Unity 8 itself, because of time constraints around moving it over to systemd.  I did that as an experiment: it's a few lines of shell script and some systemd unit files.  Surely if someone is capable of re-architecting thenentire desktop shell to rebase around Mutter or KWin or some other display manager that uses the Wayland protocol, a few lines of shell script shouldn't be an insurmountable obstacle.  Upstart is not used for launching every application.

**Comment by z3ntu** [comment]
I tried it basically when the project was renamed, created a bunch of PKGBUILDs from .debs but actually failed with libinput not having some options required for compiling mir (I think). So Ubuntu uses patched versions for that too... After this has been switched to Wayland it should be better (hopefully).

**Comment by bregma** [comment]
Yes, Ubuntu carries [this patchset](https://lists.freedesktop.org/archives/wayland-devel/2015-November/025415.html) to libinput to propagate touch attributes for touchscreens.  The patches haven't been accepted upstream because they weren't perfect in all theoretical cases, but required in Ubuntu to support actual shipping hardware in people's hands.  Perfection is the enemy of Good Enough, especially when you're actually shipping product.

**Comment by horsemanoffaith** [comment]
Shiznix- I have looked at your work on the Gentoo overlay, and it has helped me in places with Unity 7. While I wish Unity 7 would continue, I understand that things are going to move forward and that Unity 7 will more than likely go away. That's the reason why I am interest in this project. Any insights into building Unity 8 would be helpful. Is there a way I can keep in touch with you via e-mail?

**Comment by horsemanoffaith** [comment]
@bregma- you sound very knowledgeable... could I perhaps contact you (via e-mail) as well if I get stuck on something?

**Comment by shiznix** [comment]
@bregma you are incorrect on all points, but I don't have the time to prove you wrong so I wish you good luck on your merry way :)

@horsemanoffaith sure you can drop me a line on rickfharris@yahoo.com.au

**Comment by horsemanoffaith** [comment]
Thanks, Shiznix!!

**Comment by bregma** [comment]
> @bregma you are incorrect on all points, but I don't have the time to prove you wrong so I wish you good luck on your merry way :)

It's good to know this now after all the years I spent developing this stuff full-time.  Thank you for your patience and wisdom in setting me straight.

**Comment by horsemanoffaith** [comment]
@bregma- based on some of your comments in other locations here, it sounds like you developed for both Unity 7 and Unity 8. There are things I'd like to ask you, but they don't belong here. Could I contact you with those questions via e-mail?

**Comment by bregma** [comment]
@horsemanoffaith  sure see my profile

**Comment by shiznix** [comment]
@bregma Then with the greatest respect, you should have no excuse for being incorrect then :)

While I wait for my coffee to brew, here's the technical why on each point:

> Ubuntu forwards all of Debian's patches to the qtbase package and add a few more for things like tweaking appmenus. None of those patches are specific to Unity 8 or even required by Unity 8.

Unity8 desktop will not launch (Black Screen of Death) on a non-Ubuntu desktop without the Ubuntu QT5 patchset as qtmir won't work when qtcore segfaults wrt. QFactoryLoader. Cherry picking the relevant QFactoryLoader Ubuntu patch from their QT5 patchset fixes the segfault but not the BSOD. Only once the entire patchset is used will Unity8 desktop session launch. This may or may not have been fixed in Zesty (Qt-5.7.1) and you may have never come up against this developing on Ubuntu with non-vanilla QT5 so letting this one slide.

> There is nothing in the Unity 8 stack that requires LXC or LXCFS.

Are you the same bregma from https://bugs.launchpad.net/ubuntu/+source/unity8/+bug/1535058
The solution as seen in that bug is that Ubuntu now uses libpam-cgfs (which is LXCFS).

> Upstart is only required for launching Unity 8 itself, because of time constraints around moving it over to systemd. Upstart is not used for launching every application.

Here's a starting point /usr/share/upstart/sessions/application.conf
It's the Upstart job responsible for handing off to ubuntu-app-launch to launch applications.


Coffee time!

**Comment by bregma** [comment]
>     Ubuntu forwards all of Debian's patches to the qtbase package and add a few more for things like tweaking appmenus. None of those patches are specific to Unity 8 or even required by Unity 8.
> 
> Unity8 desktop will not launch (Black Screen of Death) on a non-Ubuntu desktop without the Ubuntu QT5 patchset as qtmir won't work when qtcore segfaults wrt. QFactoryLoader. Cherry picking the relevant QFactoryLoader Ubuntu patch from their QT5 patchset fixes the segfault but not the BSOD. Only once the entire patchset is used will Unity8 desktop session launch. This may or may not have been fixed in Zesty (Qt-5.7.1) and you may have never come up against this developing on Ubuntu with non-vanilla QT5 so letting this one slide.

It's true the Unity 8 devs discovered a number of bugs in Qt in each release:  the cycle involves distro-patching the libraries so work can continue and pushing the fixes upstream so they're available for everyone.  It's possible there are additional patches I missed.
 
>     There is nothing in the Unity 8 stack that requires LXC or LXCFS.
> 
> Are you the same bregma from https://bugs.launchpad.net/ubuntu/+source/unity8/+bug/1535058
> The solution as seen in that bug is that Ubuntu now uses libpam-cgfs (which is LXCFS).

Yes, I'm certainly aware there were some problems related to Upstart not handling user-session tasks properly after the kernel moved to cgroups v2, and those problem were solved by adding an additional runtime dependency on `libpam-cgfs` to the `ubuntu-app-launch` package.  That has nothing to do with LXD, it was only a clue that the problem did not occur on test systems that had LXD installed because _that_ package already had a dependency on `libpam-cgfs`.

cgfs (the cgroups pseudo-filesystem in the kernel) and lxcfs (the LXC FUSE-based filesystem) are unrelated filesystems.  They are orthogonal.

>     Upstart is only required for launching Unity 8 itself, because of time constraints around moving it over to systemd. Upstart is not used for launching every application.
> 
> Here's a starting point /usr/share/upstart/sessions/application.conf
> It's the Upstart job responsible for handing off to ubuntu-app-launch to launch applications.

I am intimate, _intimately_ familiar with how `ubuntu-app-launch` uses systemd to launch applications.

[Here](http://bazaar.launchpad.net/~indicator-applet-developers/ubuntu-app-launch/trunk.17.04/revision/303) is the commit where the last vestiges of upstart were removed.  If you prefer git, then look [here](https://github.com/unity8-team/ubuntu-app-launch/commit/76fe6e8c61226c21bf33e406caff29552f1df93f).

The file you point to does not exist in the current release.  Start by looking at Ubuntu 17.04.

**Comment by shiznix** [comment]
@bregma Thanks I see now that Upstart is now no longer required for Unity8 application launch since that commit which made the change.

Unfortunately since upgrading my Ubuntu installation from Yakkety to Zesty, Unity8 no longer launches in Zesty and gives the old BSOD that has plagued this software on and off since it's inception.
Judging by announcements from Mr. Shuttleworth I'm guessing this is the final death throes and may be a time wasting exercise to pursue this any further.
I mean if it doesn't work on Ubuntu what hope do we have packaging it for other distributions?

That said, Unity7 is still carrying on strong and continuing to use that for now.

Cheers :)

**Comment by MggMuggins** [comment]
Has anyone made any progress regarding Yunit in Arch? I'm interested in helping write PKGBUILDS and getting dependencies sorted, but does anyone have any already written that could be pushed up to a central repo so that we could consolidate our efforts?
